README.md
[![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com) [![Build Status](https://travis-ci.org/LifiPlus/node-serverless-response.svg?branch=master)](https://travis-ci.org/LifiPlus/node-serverless-response)

# node-serverless-response

If you are using the serverless framework to develop your micro service, this package can help you to generate the HTTP response payload.

# Installing

NPM

```
npm install node-serverless-response
```

Yarn

```
yarn add node-serverless-response
```

# Example


## General

```
const Response = require('node-serverless-response')
const _Response = new Response()

export.handler = async (event, context, callback) => {
  return _Response.success('Hello World!')
}
```

## Support CORS

```
const Response = require('node-serverless-response')
const _Response = new Response()

export.handler = async (event, context, callback) => {
  return _Response.success('Hello World!', true)
}
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
- None

## [1.1.1] - 2018-11-02
### Added
- ESLint
- JSON response

### Changed
- Optimize testing code
- Upgrade ESLint version

### Removed
- None

## [1.1.0] - 2018-10-29
### Added
- Add jest testing
- Travis CI

### Changed
- Fix some typo

### Removed
- None

## [1.0.0] - 2018-10-21
### Added
- First release version

### Changed
- None

### Removed
- None


__tests__/Response.test.js
const Response = require('../Response')
const _Response = new Response()

let textTemplate = {
  body: 'Test Success',
  headers: {
    'Content-Type': 'text/plain;charset=utf-8'
  }
}

let textWithCorsTemplate = {
  body: 'Test Success',
  headers: {
    'Content-Type': 'text/plain;charset=utf-8',
    'Access-Control-Allow-Credentials': true,
    'Access-Control-Allow-Origin': '*'
  }
}

let jsonTemplate = {
  body: '{"message":"Test Success"}',
  headers: {
    'Content-Type': 'application/json;charset=utf-8'
  }
}

let jsonWithCorsTemplate = {
  body: '{"message":"Test Success"}',
  headers: {
    'Content-Type': 'application/json;charset=utf-8',
    'Access-Control-Allow-Credentials': true,
    'Access-Control-Allow-Origin': '*'
  }
}

test('Build success response', () => {
  textTemplate.statusCode = 200

  expect(_Response.success('Test Success')).toMatchObject(textTemplate)
})

test('Build success response with CORS', () => {
  textWithCorsTemplate.statusCode = 200

  expect(_Response.success('Test Success', true)).toMatchObject(textWithCorsTemplate)
})

test('Build success JSON response', () => {
  jsonTemplate.statusCode = 200

  expect(_Response.success({message: 'Test Success'})).toMatchObject(jsonTemplate)
})

test('Build success JSON response with CORS', () => {
  jsonWithCorsTemplate.statusCode = 200

  expect(_Response.success({message: 'Test Success'}, true)).toMatchObject(jsonWithCorsTemplate)
})

test('Build bad request response', () => {
  textTemplate.statusCode = 400

  expect(_Response.badRequest('Test Success')).toMatchObject(textTemplate)
})

test('Build bad request response with CORS', () => {
  textWithCorsTemplate.statusCode = 400

  expect(_Response.badRequest('Test Success', true)).toMatchObject(textWithCorsTemplate)
})

test('Build bad request JSON response', () => {
  jsonTemplate.statusCode = 400

  expect(_Response.badRequest({message: 'Test Success'})).toMatchObject(jsonTemplate)
})

test('Build bad request JSON response with CORS', () => {
  jsonWithCorsTemplate.statusCode = 400

  expect(_Response.badRequest({message: 'Test Success'}, true)).toMatchObject(jsonWithCorsTemplate)
})

test('Build unauthorized response', () => {
  textTemplate.statusCode = 401
  expect(_Response.unauthorized('Test Success')).toMatchObject(textTemplate)
})

test('Build unauthorized response with CORS', () => {
  textWithCorsTemplate.statusCode = 401

  expect(_Response.unauthorized('Test Success', true)).toMatchObject(textWithCorsTemplate)
})

test('Build unauthorized JSON response', () => {
  jsonTemplate.statusCode = 401

  expect(_Response.unauthorized({message: 'Test Success'})).toMatchObject(jsonTemplate)
})

test('Build unauthorized JSON response with CORS', () => {
  jsonWithCorsTemplate.statusCode = 401

  expect(_Response.unauthorized({message: 'Test Success'}, true)).toMatchObject(jsonWithCorsTemplate)
})

test('Build forbidden response', () => {
  textTemplate.statusCode = 403

  expect(_Response.forbidden('Test Success')).toMatchObject(textTemplate)
})

test('Build forbidden response with CORS', () => {
  textWithCorsTemplate.statusCode = 403

  expect(_Response.forbidden('Test Success', true)).toMatchObject(textWithCorsTemplate)
})

test('Build forbidden JSON response', () => {
  jsonTemplate.statusCode = 403

  expect(_Response.forbidden({message: 'Test Success'})).toMatchObject(jsonTemplate)
})

test('Build forbidden JSON response with CORS', () => {
  jsonWithCorsTemplate.statusCode = 403

  expect(_Response.forbidden({message: 'Test Success'}, true)).toMatchObject(jsonWithCorsTemplate)
})

test('Build not found response', () => {
  textTemplate.statusCode = 404
  
  expect(_Response.notFound('Test Success')).toMatchObject(textTemplate)
})

test('Build not found response with CORS', () => {
  textWithCorsTemplate.statusCode = 404

  expect(_Response.notFound('Test Success', true)).toMatchObject(textWithCorsTemplate)
})

test('Build not found JSON response', () => {
  jsonTemplate.statusCode = 404

  expect(_Response.notFound({message: 'Test Success'})).toMatchObject(jsonTemplate)
})

test('Build not found JSON response with CORS', () => {
  jsonWithCorsTemplate.statusCode = 404

  expect(_Response.notFound({message: 'Test Success'}, true)).toMatchObject(jsonWithCorsTemplate)
})

test('Build method not allowed response', () => {
  textTemplate.statusCode = 405

  expect(_Response.methodNotAllowed('Test Success')).toMatchObject(textTemplate)
})

test('Build method not allowed response with CORS', () => {
  textWithCorsTemplate.statusCode = 405

  expect(_Response.methodNotAllowed('Test Success', true)).toMatchObject(textWithCorsTemplate)
})

test('Build method not allowed JSON response', () => {
  jsonTemplate.statusCode = 405

  expect(_Response.methodNotAllowed({message: 'Test Success'})).toMatchObject(jsonTemplate)
})

test('Build method not allowed JSON response with CORS', () => {
  jsonWithCorsTemplate.statusCode = 405

  expect(_Response.methodNotAllowed({message: 'Test Success'}, true)).toMatchObject(jsonWithCorsTemplate)
})


jest.config.js
// For a detailed explanation regarding each configuration property, visit:
// https://jestjs.io/docs/en/configuration.html

module.exports = {
  // All imported modules in your tests should be mocked automatically
  // automock: false,

  // Stop running tests after the first failure
  // bail: false,

  // Respect "browser" field in package.json when resolving modules
  // browser: false,

  // The directory where Jest should store its cached dependency information
  // cacheDirectory: "/var/folders/m0/_9179l9507jf8r61173wrvg40000gn/T/jest_dx",

  // Automatically clear mock calls and instances between every test
  clearMocks: true,

  // Indicates whether the coverage information should be collected while executing the test
  // collectCoverage: false,

  // An array of glob patterns indicating a set of files for which coverage information should be collected
  // collectCoverageFrom: null,

  // The directory where Jest should output its coverage files
  coverageDirectory: "coverage",

  // An array of regexp pattern strings used to skip coverage collection
  // coveragePathIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // A list of reporter names that Jest uses when writing coverage reports
  // coverageReporters: [
  //   "json",
  //   "text",
  //   "lcov",
  //   "clover"
  // ],

  // An object that configures minimum threshold enforcement for coverage results
  // coverageThreshold: null,

  // Make calling deprecated APIs throw helpful error messages
  // errorOnDeprecated: false,

  // Force coverage collection from ignored files usin a array of glob patterns
  // forceCoverageMatch: [],

  // A path to a module which exports an async function that is triggered once before all test suites
  // globalSetup: null,

  // A path to a module which exports an async function that is triggered once after all test suites
  // globalTeardown: null,

  // A set of global variables that need to be available in all test environments
  // globals: {},

  // An array of directory names to be searched recursively up from the requiring module's location
  // moduleDirectories: [
  //   "node_modules"
  // ],

  // An array of file extensions your modules use
  // moduleFileExtensions: [
  //   "js",
  //   "json",
  //   "jsx",
  //   "node"
  // ],

  // A map from regular expressions to module names that allow to stub out resources with a single module
  // moduleNameMapper: {},

  // An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
  // modulePathIgnorePatterns: [],

  // Activates notifications for test results
  // notify: false,

  // An enum that specifies notification mode. Requires { notify: true }
  // notifyMode: "always",

  // A preset that is used as a base for Jest's configuration
  // preset: null,

  // Run tests from one or more projects
  // projects: null,

  // Use this configuration option to add custom reporters to Jest
  // reporters: undefined,

  // Automatically reset mock state between every test
  // resetMocks: false,

  // Reset the module registry before running each individual test
  // resetModules: false,

  // A path to a custom resolver
  // resolver: null,

  // Automatically restore mock state between every test
  // restoreMocks: false,

  // The root directory that Jest should scan for tests and modules within
  // rootDir: null,

  // A list of paths to directories that Jest should use to search for files in
  // roots: [
  //   "<rootDir>"
  // ],

  // Allows you to use a custom runner instead of Jest's default test runner
  // runner: "jest-runner",

  // The paths to modules that run some code to configure or set up the testing environment before each test
  // setupFiles: [],

  // The path to a module that runs some code to configure or set up the testing framework before each test
  // setupTestFrameworkScriptFile: null,

  // A list of paths to snapshot serializer modules Jest should use for snapshot testing
  // snapshotSerializers: [],

  // The test environment that will be used for testing
  testEnvironment: "node",

  // Options that will be passed to the testEnvironment
  // testEnvironmentOptions: {},

  // Adds a location field to test results
  // testLocationInResults: false,

  // The glob patterns Jest uses to detect test files
  // testMatch: [
  //   "**/__tests__/**/*.js?(x)",
  //   "**/?(*.)+(spec|test).js?(x)"
  // ],

  // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
  // testPathIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // The regexp pattern Jest uses to detect test files
  // testRegex: "",

  // This option allows the use of a custom results processor
  // testResultsProcessor: null,

  // This option allows use of a custom test runner
  // testRunner: "jasmine2",

  // This option sets the URL for the jsdom environment. It is reflected in properties such as location.href
  // testURL: "http://localhost",

  // Setting this value to "fake" allows the use of fake timers for functions such as "setTimeout"
  // timers: "real",

  // A map from regular expressions to paths to transformers
  // transform: null,

  // An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
  // transformIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
  // unmockedModulePathPatterns: undefined,

  // Indicates whether each individual test should be reported during the run
  // verbose: null,

  // An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
  // watchPathIgnorePatterns: [],

  // Whether to use watchman for file crawling
  // watchman: true,
};


Response/index.js
'use strict'

/**
 * Generate common responses
 */
class Response {
  /**
   * Build HTTP response
   * @param {Number} statusCode - Response HTTP status code
   * @param {Object|String} body - Response payload
   * @param {Boolean} isCors - Is support CORS or not
   * @return {Object} - Response Payload
   * @example
   * {
   *   statusCode: 200,
   *   headers: {
   *     'Access-Control-Allow-Origin': '*',
   *     'Access-Control-Allow-Credentials': true
   *   },
   *   body: '{"message": "Success"}'
   * }
   */
  _buildResponse (statusCode, body, isCors = false) {
    let headers = {}

    if (body.constructor === Object) {
      headers['Content-Type'] = 'application/json;charset=utf-8'
      body = JSON.stringify(body)
    } else {
      headers['Content-Type'] = 'text/plain;charset=utf-8'
    }

    if (isCors) {
      headers['Access-Control-Allow-Origin'] = '*'
      headers['Access-Control-Allow-Credentials'] = true
    }

    return {
      statusCode,
      headers: headers,
      body: body
    }
  }

  /**
   * Request success
   * @param {Object|String} body - Response payload
   * @param {Boolean} isCors - Is support CORS or not
   * @return {Object|String} - Please see {@link _buildResponse} for more detail
   */
  success (body, isCors = false) {
    return this._buildResponse(200, body, isCors)
  }

  /**
   * Bad request
   * @param {Object|String} body - Response payload
   * @param {Boolean} isCors - Is support CORS or not
   * @return {Object|String} - Please see {@link _buildResponse} for more detail
   */
  badRequest (body, isCors = false) {
    return this._buildResponse(400, body, isCors)
  }

  /**
   * Request unauthorized
   * @param {Object|String} body - Response payload
   * @param {Boolean} isCors - Is support CORS or not
   * @return {Object|String} - Please see {@link _buildResponse} for more detail
   */
  unauthorized (body, isCors = false) {
    return this._buildResponse(401, body, isCors)
  }

  /**
   * Request forbidden
   * @param {Object|String} body - Response payload
   * @param {Boolean} isCors - Is support CORS or not
   * @return {Object|String} - Please see {@link _buildResponse} for more detail
   */
  forbidden (body, isCors = false) {
    return this._buildResponse(403, body, isCors)
  }

  /**
   * Something not found
   * @param {Object|String} body - Response payload
   * @param {Boolean} isCors - Is support CORS or not
   * @return {Object|String} - Please see {@link _buildResponse} for more detail
   */
  notFound (body, isCors = false) {
    return this._buildResponse(404, body, isCors)
  }

  /**
   * Request methos not allowed
   * @param {Object|String} body - Response payload
   * @param {Boolean} isCors - Is support CORS or not
   * @return {Object|String} - Please see {@link _buildResponse} for more detail
   */
  methodNotAllowed (body, isCors = false) {
    return this._buildResponse(405, body, isCors)
  }

  /**
   * Request too many
   * @param {Object|String} body - Response payload
   * @param {Boolean} isCors - Is support CORS or not
   * @return {Object|String} - Please see {@link _buildResponse} for more detail
   */
  tooManyRequests (body, isCors = false) {
    return this._buildResponse(429, body, isCors)
  }

  /**
   * Internal error
   * @param {Object|String} body - Response payload
   * @param {Boolean} isCors - Is support CORS or not
   * @return {Object|String} - Please see {@link _buildResponse} for more detail
   */
  internalError (body, isCors = false) {
    return this._buildResponse(500, body, isCors)
  }
}

module.exports = Response
